src/Servers/Kestrel/Core/src/Internal/SniOptionsSelector.cs

```diff
@@ -0,0 +1,223 @@
+            sniOptions ??= _wildcardOptions;
+
+            if (sniOptions is null)
+            {
+                if (serverName is null)
+                {
+                    // There was no ALPN
+                    throw new AuthenticationException(CoreStrings.FormatSniNotConfiguredToAllowNoServerName(_endpointName));
+                }
+                else
+                {
+                    throw new AuthenticationException(CoreStrings.FormatSniNotConfiguredForServerName(serverName, _endpointName));
+                }
+            }
+
+            connection.Features.Set(new HttpProtocolsFeature(sniOptions.HttpProtocols));
+
+            var sslOptions = sniOptions.SslOptions;
+
+            if (sslOptions.ServerCertificate is null)
+            {
+                Debug.Assert(_fallbackServerCertificateSelector != null,
+                    "The cached SniOptions ServerCertificate can only be null if there's a fallback certificate selector.");
+
+                // If a ServerCertificateSelector doesn't return a cert, HttpsConnectionMiddleware doesn't fallback to the ServerCertificate.
+                sslOptions = CloneSslOptions(sslOptions);
+                var fallbackCertificate = _fallbackServerCertificateSelector(connection, serverName);
+
+                if (fallbackCertificate != null)
+                {
+                    HttpsConnectionMiddleware.EnsureCertificateIsAllowedForServerAuth(fallbackCertificate);
+                }
+
+                sslOptions.ServerCertificate = fallbackCertificate;
+            }
+
+            if (_onAuthenticateCallback != null)
+            {
+                // From doc comments: "This is called after all of the other settings have already been applied."
+                sslOptions = CloneSslOptions(sslOptions);
+                _onAuthenticateCallback(connection, sslOptions);
+            }
+
+            return sslOptions;
+        }
+
+        public static ValueTask<SslServerAuthenticationOptions> OptionsCallback(ConnectionContext connection, SslStream stream, SslClientHelloInfo clientHelloInfo, object state, CancellationToken cancellationToken)
+        {
+            var sniOptionsSelector = (SniOptionsSelector)state;
+            var options = sniOptionsSelector.GetOptions(connection, clientHelloInfo.ServerName);
+            return new ValueTask<SslServerAuthenticationOptions>(options);
+        }
+
+        // TODO: Reflection based test to ensure we clone everything!
+        // This won't catch issues related to mutable subproperties, but the existing subproperties look like they're mostly immutable.
+        // The exception are the ApplicationProtocols list which we clone and the ServerCertificate because of methods like Import() and Reset() :(
+        internal static SslServerAuthenticationOptions CloneSslOptions(SslServerAuthenticationOptions sslOptions) =>
+            new SslServerAuthenticationOptions
```

Comment:
Should we ask the runtime to add a clone method on SslServerAuthenticationOptions?

Reply:
Maybe. I added the test. I don't thing it's a huge deal either way.